messaging/src/test/java/org/springframework/security/messaging/context/SecurityContextChannelInterceptorTests.java

```diff
@@ -175,6 +169,24 @@ public void beforeHandleUserSet() {
 		assertThat(SecurityContextHolder.getContext().getAuthentication()).isSameAs(this.authentication);
 	}
 
+	@Test
+	public void postReceiveUserSet() {
+		this.messageBuilder.setHeader(SimpMessageHeaderAccessor.USER_HEADER, this.authentication);
+		this.interceptor.postReceive(this.messageBuilder.build(), this.channel);
+		assertThat(SecurityContextHolder.getContext().getAuthentication()).isSameAs(this.authentication);
+	}
+
+	@Test
+	public void authenticationIsPropagatedFromPreSendToPostReceive() {
```

Comment:
I don't think I'm understanding this test. What is the real-world scenario that it is simulating?

Reply:
The use-case is exactly what the change in this PR is needed. We put a message into a `Queue` and we don't know which thread, where and when we going to pull from that queue. In preSend we store auth ctx into message header and in the `postReceive` we do restore that ctx into the current thread which has just called `receive` on a queue.

All of that logic is really covered by the `SecurityContextChannelInterceptor` with a small adjustments to a missed functionality for `postReceive()`.

Reply:
The thing I'm not understanding is why would you want the `SecurityContextHolder` to be cleared after `afterMessageHandled` is called? 

The following flow feels more natural: `preSend` saves the current security context and sets a new one according to the message contents, then `cleanup` restores what it was before.